client/src/main/kotlin/io/spine/examples/pingh/client/LoginFlow.kt

```diff
@@ -0,0 +1,309 @@
+ * first enter a username and obtain a user code, then verify the login.
+ *
+ * @param client enables interaction with the Pingh server.
+ * @param session the information about the current user session.
+ */
+public class LoginFlow internal constructor(
+    private val client: DesktopClient,
+    private val session: MutableStateFlow<UserSession?>
+) {
+
+    /**
+     * Internal data of the login process.
+     */
+    private val context = LoginContext()
+
+    /**
+     * The current stage of the login process.
+     */
+    @Suppress("MemberVisibilityCanBePrivate" /* Accessed from `desktop` module. */)
+    public val stage: MutableStateFlow<LoginStageType>
+        get() = context.stage
+
+    /**
+     * Initiates the stage of the login process where the user must enter their username
+     * and receive a user code.
+     *
+     * @throws IllegalStateException if the current login stage is not `EnterUsername`.
+     */
+    public fun askForUsername(): EnterUsername {
+        check(stage.value == EnterUsername::class) {
+            "The current login stage must be `EnterUsername`."
+        }
+        return EnterUsername(client, session, context)
+    }
+
+    /**
+     * Initiates the stage of the login process where the user must verify their login to GitHub.
+     *
+     * @throws IllegalStateException if the current login stage is not `VerifyLogin`.
+     */
+    public fun verifyLogin(): VerifyLogin {
+        check(stage.value == VerifyLogin::class) {
+            "The current login stage must be `VerifyLogin`."
+        }
+        return VerifyLogin(client, session, context)
+    }
+}
+
+/**
+ * Stages of login process.
+ */
+public interface LoginStage
+
+/**
+ * Type of the login process stage.
+ */
+public typealias LoginStageType = KClass<out LoginStage>
+
+/**
+ * Internal data of the login process.
+ */
+internal class LoginContext {
+
+    /**
+     * The current stage of the login process.
+     */
+    internal val stage: MutableStateFlow<LoginStageType> = MutableStateFlow(EnterUsername::class)
+
+    /**
+     * Stores the event received after the user enters their name.
+     *
+     * This is required to initialize the verification stage.
+     */
+    internal var userCodeReceived: UserCodeReceived? = null
+}
+
+/**
+ * Manages the stage of the login process where the user must enter their username
+ * will receive a user code in return.
+ *
+ * @param client enables interaction with the Pingh server.
+ * @param session the information about the current user session.
+ * @param context the login process data.
+ */
+public class EnterUsername internal constructor(
+    private val client: DesktopClient,
+    private val session: MutableStateFlow<UserSession?>,
+    private val context: LoginContext
+) : LoginStage {
+
+    /**
+     * Starts the login process and requests `UserCode`.
+     */
+    public fun requestUserCode(
+        username: Username,
+        onSuccess: (event: UserCodeReceived) -> Unit = {}
```

Comment:
I don't like the idea of wiring the stages in the "client's" code. This flow defines the steps. It also must define allowed transitions.

Right now, users of this API must guess which stage is the next one (which is fine only in case there's only two stages). Otherwise, they will get an `IllegalStateException`, at run-time.